// vayu/models/Board.js
var util = require('util')
    , VMB = require('./VayuModelBase.js')
    , moment = require('moment');

function Board()
{
    VMB.call(this);
}

util.inherits(Board, VMB);

var p = Board.prototype;

p.getBoards = function (userId, success, fail)
{
    this.Model.find({u: userId}, null, {sort: {n: 1}}).exec(function (err, doc) {
        if (err) {
            fail(err);
        } else {
            success(doc);
        }
    });
}

p.addBoard = function (userId, lid, name, success, fail)
{
    //TODO: handle multiple persons
    var b = new this.Model({
        lid: lid,
        u: userId,
        n: name
    });
    this.saveModel(b, success, fail);
}

p.updateBoard = function(userId, id, lid, name, cards, success, fail) {
    this.Model.update({
        u: userId,
        _id: id
    }, {
        id: lid,
        n: name,
        cards: cards
    }, function(err, numAff, response) {
        if(err) {
            fail(err);
        } else {
            // update only one word
            //TODO: error tracking
            if(numAff == 1) {
                success();
            }else{
                console.log("error saving board...");
                console.log(err);
                fail(err);
            }
        }
    });
}

p.deleteAll = function(userId, success, fail) {
    this.Model.remove({
        u: userId
    }, function(err)
    {
        console.log(err);
        if(err)
        {
            fail(err);
        }else{
            success();
        }
    });
}

var board = new Board();

board.init('Board', 'board', {
    //id from frontend (local)
    lid: String,
    //user id
    u: { type: board.Schema.ObjectId, ref: "User" },
    //name
    n: String,
    //cards
    //TODO: shorten the name
    cards: [],
    //created date
    cd: { "type": Date, "default": Date.now },
    //date for displaying on the frontend
});

module.exports = exports = board;
