// vayu/models/User.js
var util = require('util');
var VMB = require('./VayuModelBase.js');

function User() {
    VMB.call(this);
    this.init('User', 'user', {
        //fname
        f: String,
        //lname
        l: String,
        //gender
        g: String,
        //email
        e: String,
        //dob
        d: Date,
        //pob
        p: String,
        //profile picture
        pp: String,
        //time zone
        t: String,

        //TODO: move these to a better place and into a single object, since this is social data!
        //Facebook ID
        fb: String,
        //Facebook access token
        fb_tk: String,
        //Facebook refresh token
        fb_re: String,
        //Other data form facebook
        fb_o: this.Schema.Types.Mixed
    });
}

util.inherits(User, VMB);

var p = User.prototype;

//utility methods
p.findByEmail = function (email, success, fail) {
    this.Model.findOne({
        e: email
    }, function (err, doc) {
        if (err) {
            fail(err);
        } else {
            success(doc);
        }
    });
};

p.findByFBID = function (id, success, fail) {
    this.Model.findOne({
        fb: id
    }, function (err, doc) {
        if (err) {
            fail(err);
        } else {
            success(doc);
        }
    });
};

p.findByID = function (id, success, fail) {
    this.Model.findOne({
        _id: id
    }, function (err, doc) {
        if (err) {
            fail(err);
        } else {
            success(doc);
        }
    });
};

p.registerUserFB = function (fname, lname, gender, email, fbID, accessToken, refreshToken, otherData, success, fail) {
    var u = new this.Model({
        f: fname,
        l: lname,
        g: gender,
        e: email,
        fb: fbID,
        fb_tk: accessToken,
        fb_re: refreshToken,
        fb_o: otherData
    });
    this.saveModel(u, success, fail);
};

p.findOrRegisterFB = function (accessToken, refreshToken, obj, success, fail) {
    var root = this;
    this.Model.findOne({
        fb: obj.id
    }, function (err, doc) {
        if (err) {
            fail(err);
        } else {
            if (doc !== null) {
                success(doc);
            } else {
                root.registerUserFB(obj.first_name, obj.last_name, obj.gender === 'male' ? 'm' : 'f', obj.email, obj.id, accessToken, refreshToken, obj, success, fail);
            }
        }
    });
};

p.registerUser = function (fname, lname, gender, email, dob, pob, success, fail) {
    var u = new this.Model({
        f: fname,
        l: lname,
        g: gender,
        e: email,
        d: dob,
        p: pob
    });
    this.saveModel(u, success, fail);
};

//virtuals
//options - read options, safe options, strict, etc

module.exports = exports = new User();
