// vayu/vayu.js
var Vayu = function () {
	
    var passport = require('passport'),
        FacebookStrategy = require('passport-facebook').Strategy,
        nconf = require('nconf'),
        mongoose = require('mongoose');
    
    var fb_cb_success = "";
    var fb_cb_error = "";
    
    var _user;
    var _card;

    var _app;
    
    passport.serializeUser(function (user, done) {
        done(null, user._id);
    });

    passport.deserializeUser(function (id, done) {
        _user.findByID(id, function (doc) {
            done(null, doc);
        }, function (err) {
            done(err);
        });
    });
    
    function init(app)
    {
        _app = app;

        //configuration
        nconf.argv().env();
        
        //initialize models
        _user = require('./models/User.js');
        _board = require('./models/Board.js');
        _card = require('./models/Card.js');

        setupRoutes();
    }

    var setupRoutes = function() {
        passport.use(new FacebookStrategy({
            clientID: nconf.get('VAYU_FB_APP_ID'),
            clientSecret: nconf.get('VAYU_FB_APP_SECRET'),
            callbackURL: nconf.get('VAYU_FB_CALLBACK')
        }, addOrGetUser));
        
        mongoose.connect(nconf.get('VAYU_DB'));
        
        //routes
        _app.get("/connect/fb/cb", passport.authenticate('facebook', {
            successRedirect: "/me",
            failureRedirect: "/error.html"
        }));
        
        _app.get("/connect/fb", passport.authenticate('facebook', {
            scope: ['email']
        }));
        
        _app.get("/me", function(req, res){
            //FIXME: not enough to just check user
            if(req.user)
            {
                res.render('me', {
                    title: req.user.f
                });
            }else{
                res.redirect("/index.html");
            }
        });

        _app.get("/boards", boardsGetHandler);
        _app.post("/boards", boardsPostHandler);
        _app.put("/boards", boardsPutHandler);

        _app.get("/cards", cardsGetHandler);
        _app.post("/cards", cardsPostHandler)
        _app.put("/cards", cardsPutHandler)
    }
    
    // routes handlers
    var addOrGetUser = function (accessToken, refreshToken, profile, done) {
        _user.findOrRegisterFB(accessToken, refreshToken, profile._json, function (doc) {
            done(null, doc);
        }, function (err) {
            //TODO: different responses based on error!
            done(err);
        });
    };

    var boardsGetHandler = function (req, res) {
        getBoards(req.user._id, function (boards) {
            res.json(boards);
        }, function (err) {
            res.send(500);
        });
    }

    var boardsPostHandler = function(req, res) {
        addBoard(req.user._id, req.body.lid, req.body.name, function(board) {
            res.json(board);
        }, function(err) {
            res.send(500);
        });
    }

    var boardsPutHandler = function(req, res) {
        // TODO: really needed to create separate array?
        var c = [];
        var cards = req.body.cards;
        var len = cards.length;

        for (var i=0; i < len; i++) {
            c.push(cards[i]);
        }

        updateBoard(req.user._id, req.body._id, req.body.lid, req.body.name, c, function() {
            //NOTE: some backbonejs thing here...expects a "200" response as string
            res.send(200, "200");
        }, function(err) {
            res.send(500);
        });
    }

    var cardsGetHandler = function (req, res) {
        getCards(req.user._id, req.query.bid, function (cards) {
            res.json(cards);
        }, function (err) {
            res.send(500);
        });
    }

    var cardsPostHandler = function(req, res) {
        addCard(req.user._id, req.body.bid, req.body.lid, req.body.name, function(card) {
            res.json(card);
        }, function(err) {
            res.send(500);
        });
    }

    var cardsPutHandler = function(req, res) {
        var c = [];
        var cards = req.body.cards;
        var len = cards.length;

        for (var i=0; i < len; i++) {
            c.push(cards[i]);
        }

        updateCard(req.user._id, req.body._id, req.body.lid, req.body.name, c, function() {
            res.send(200);
        }, function(err) {
            res.send(500);
        });
    }

    // model interactions
    // TODO: move to controllers
    var getBoards = function (userId, success, fail) {
        //TODO: segreate based on date range
        _board.getBoards(userId, success, fail);
    };

    var addBoard = function(userId, lid, name, success, fail) {
        _board.addBoard(userId, lid, name, success, fail);
    }

    var updateBoard = function(userId, id, lid, name, cards, success, fail) {
        _board.updateBoard(userId, id, lid, name, cards, success, fail);
    }

    var boardsGetHandler = function (req, res) {
        getBoards(req.user._id, function (boards) {
            res.json(boards);
        }, function (err) {
            res.send(500);
        });
    }

    var getCards = function (userId, bid, success, fail) {
        _card.getCards(userId, bid, success, fail);
    };

    var addCard = function(userId, bid, lid, name, success, fail) {
        _card.addCard(userId, bid, lid, name, success, fail);
    }

    var updateCard = function(userId, id, lid, name, cards, success, fail) {
        _board.updateBoard(userId, id, lid, name, cards, success, fail);
    }
    
    return {
        init: init
    };
};

exports = module.exports = Vayu();
