// vayu/vayu.js
var Vayu = function () {
	
    var passport = require('passport'),
        FacebookStrategy = require('passport-facebook').Strategy,
        nconf = require('nconf'),
        mongoose = require('mongoose');
    
    var fb_cb_success = "";
    var fb_cb_error = "";
    
    var _user;
    
    passport.serializeUser(function (user, done) {
        done(null, user._id);
    });

    passport.deserializeUser(function (id, done) {
        _user.findByID(id, function (doc) {
            done(null, doc);
        }, function (err) {
            done(err);
        });
    });
    
    function init(app)
    {
        //configuration
        nconf.argv().env();
        
        //initialize models
        _user = require('./models/User.js');
        
        passport.use(new FacebookStrategy({
            clientID: nconf.get('VAYU_FB_APP_ID'),
            clientSecret: nconf.get('VAYU_FB_APP_SECRET'),
            callbackURL: nconf.get('VAYU_FB_CALLBACK')
        }, addOrGetUser));
        
        mongoose.connect(nconf.get('VAYU_DB'));
        
        //routes
        app.get("/connect/fb/cb", passport.authenticate('facebook', {
            successRedirect: "/me",
            failureRedirect: "/error.html"
        }));
        
        app.get("/connect/fb", passport.authenticate('facebook', {
            scope: ['email']
        }));
        
        app.get("/me", function(req, res){
            //FIXME: not enough to just check user
            if(req.user)
            {
                res.render('me', {
                    title: req.user.f
                });
            }else{
                res.redirect("/index.html");
            }
        });
    }
    
    var addOrGetUser = function (accessToken, refreshToken, profile, done) {

        _user.findOrRegisterFB(accessToken, refreshToken, profile._json, function (doc) {
            done(null, doc);
        }, function (err) {
            //TODO: different responses based on error!
            done(err);
        });
    };
    
    return {
        init: init
    };
};

exports = module.exports = Vayu();
